packages/lit-dev-content/src/global/dsd-polyfill.ts

```diff
@@ -0,0 +1,11 @@
+(async () => {
+  if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
+    const {hydrateShadowRoots} = await import(
```

Comment:
Since this is an async operation, it's not going to block the loading of component definitions (even though the _module_ is guaranteed to execute before component scripts do -- that's not sufficient here).

I believe that means there's a race condition where a component could upgrade before hydration support was installed, so it wouldn't hydrate.

This is why the example strategy at https://github.com/lit/lit/tree/main/packages/labs/eleventy-plugin-lit#example-bootup-strategy moves component modules into dynamic `import` scripts after this promise has resolved.

Reply:
(Minor correction, I said "hydration support" above, but I should have said "the declarative shadow dom polyfill". A similar timing issue exists though -- the polyfill must have installed before any components upgrade.)

I think we'd need to define a `components.js` or similar module for every page, instead of individual component script tags, so that we can defer component imports until the polyfill has run:

```ts
import {installDsdPolyfillIfNeeded} from '../dsd-polyfill.js';
(async() {
  await installDsdPolyfillIfNeeded();
  import('../components/foo.js');
  import('../components/bar.js');
})();
```

*Or*, another approach would be to always import the polyfill (maybe even inline it), and do the check + install synchronously. This is actually probably the simpler way to go, since it just requires that you put the polyfill `<script>` tag before any component definitions. The downside is that we always download the polyfill even on Chrome -- but it's so small that this probably doesn't matter (and maybe we can make it even smaller). I think I'm leaning towards this approach...

Reply:
TBH I feel like the underlying issue here is that the _right way_<sup>(tm)</sup> is bad DX as you have to define a new entrypoint in the rollup config after you create your new `.ts` file. How about we make a subdir in our src directory called something like `entrypoints` or `/src/pages/entrypoints` that rollup just does an `fs` call and automatically create a new bundle for files in that directory.

That should enable us to do the correct method more easily.

Reply:
I think the "always load the polyfill" approach is the simplest, especially for now, and maybe forever. The declarative shadow dom polyfill is really small. So if we just put its script tag first, nothing much else needs to change with our bundling strategy.